src/server/conn/auto.rs

```diff
@@ -0,0 +1,409 @@
+//! Http1 or Http2 connection.
+
+use crate::{common::rewind::Rewind, rt::tokio_executor::TokioExecutor, Result};
+use bytes::Bytes;
+use http::{Request, Response};
+use http_body::Body;
+use hyper::{
+    body::Incoming,
+    rt::Timer,
+    server::conn::{http1, http2},
+    service::Service,
+};
+use std::{error::Error as StdError, marker::Unpin, time::Duration};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite};
+
+const H2_PREFACE: &[u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+/// Http1 or Http2 connection builder.
+pub struct Builder {
+    http1: http1::Builder,
+    http2: http2::Builder<TokioExecutor>,
+}
+
+impl Builder {
+    /// Create a new auto connection builder.
+    pub fn new() -> Self {
+        Self {
+            http1: http1::Builder::new(),
+            http2: http2::Builder::new(TokioExecutor::new()),
```

Comment:
Could this constructor take an `http1::Builder` and `http2::Builder`, then the user can create two configurations (and set up executors/timer-handlers etc.) then the AutoConnection simply handles the protocol selection then defers to the appropriately configured HTTP stack.

As it is the AutoConnection has to expose forwarded configuration hooks for each function on the two builders and, if it misses one (e.g. because a new one is added in the hyper upstream) that configuration isn't usable until a partner change is made here.

(An alternative might be to expose `fn get_http1(&mut self) -> &mut http1::Builder` but it feels simpler to just make an HTTP1 configuration and an HTTP2 configuration then merge them together).

Reply:
Maybe combine both approaches somehow? Have a nice builder API and allow passing http1 and http2 builders.